components/server/src/workspace/gitpod-server-impl.ts

```diff
@@ -1929,6 +1932,61 @@ export class GitpodServerImpl implements GitpodServerWithTracing, Disposable {
         await this.userDB.deleteEnvVar(envvar);
     }
 
+    async hasSSHPublicKey(ctx: TraceContext): Promise<boolean> {
+        const user = this.checkUser("hasSSHPublicKey");
+        return this.userDB.hasSSHPublicKey(user.id);
+    }
+
+    async getSSHPublicKeys(ctx: TraceContext): Promise<UserSSHPublicKeyValue[]> {
+        const user = this.checkUser("getSSHPublicKeys");
+        const list = await this.userDB.getSSHPublicKeys(user.id);
+        return list.map((e) => ({
+            id: e.id,
+            name: e.name,
+            fingerprint: e.fingerprint,
+            creationTime: e.creationTime,
+            lastUsedTime: e.lastUsedTime,
+        }));
+    }
+
+    async addSSHPublicKey(ctx: TraceContext, value: SSHPublicKeyValue): Promise<UserSSHPublicKeyValue> {
+        const user = this.checkUser("addSSHPublicKey");
+        const data = await this.userDB.addSSHPublicKey(user.id, value);
+        this.updateSSHKeysForRegularRunningInstances(ctx, user.id);
+        return {
+            id: data.id,
+            name: data.name,
+            fingerprint: data.fingerprint,
+            creationTime: data.creationTime,
+            lastUsedTime: data.lastUsedTime,
+        };
+    }
+
+    async deleteSSHPublicKey(ctx: TraceContext, id: string): Promise<void> {
+        const user = this.checkUser("deleteSSHPublicKey");
+        await this.userDB.deleteSSHPublicKey(user.id, id);
+        this.updateSSHKeysForRegularRunningInstances(ctx, user.id);
+        return;
+    }
+
+    async updateSSHKeysForRegularRunningInstances(ctx: TraceContext, userId: string) {
```

Comment:
This should not be publicly exposed, right?

```suggestion
    protected async updateSSHKeysForRegularRunningInstances(ctx: TraceContext, userId: string) {
```